integrations/apify/src/setup/handleCrawlerCompleted.ts

```diff
@@ -0,0 +1,231 @@
+    let continuationState
+    try {
+      continuationState = await client.getState({
+        type: 'integration',
+        id: ctx.integrationId,
+        name: 'syncContinuation',
+      })
+    } catch (error) {
+      continuationState = null
+    }
+
+    let kbId: string
+    let startOffset = 0
+    let isContinuation = false
+
+    if (continuationState?.state?.payload?.runId === runId) {
+      isContinuation = true
+      const continuation = continuationState.state.payload
+      kbId = continuation.kbId
+      startOffset = continuation.nextOffset
+      logger.forBot().info(`▶ Continuation from offset ${startOffset}`)
+
+      try {
+        await client.setState({
+          type: 'integration',
+          id: ctx.integrationId,
+          name: 'syncContinuation',
+          payload: null,
+        })
+      } catch (error) {
+        logger.forBot().error(`Failed to clear sync continuation state: ${error}`)
+        logger.forBot().error(`Webhook will be retried by Apify`)
+        throw new RuntimeError('Failed to clear sync continuation state')
+      }
+    } else {
+      let mapping
+      try {
+        mapping = await client.getState({
+          type: 'integration',
+          id: ctx.integrationId,
+          name: 'apifyRunMappings',
+        })
+      } catch (error) {
+        logger.forBot().error(`Failed to fetch run mapping for ${runId}: ${error}`)
+        logger.forBot().error(`Webhook will be retried by Apify`)
+        throw new RuntimeError('Failed to fetch run mapping')
+      }
+
+      const mapPayload = mapping?.state?.payload
+      const mappedKbId = mapPayload?.[runId]
+
+      if (!mappedKbId) {
+        const errorMsg = `No kbId mapping found for run ${runId}. Run may not have been started through this integration.`
+        logger.forBot().error(errorMsg)
+        throw new RuntimeError(errorMsg)
+      }
+
+      kbId = mappedKbId
+      logger.forBot().info(`▶ Starting sync: ${runId}`)
+    }
+
+    // check for duplicate webhooks
+    if (!isContinuation) {
+      let syncLock
+      try {
+        syncLock = await client.getState({
+          type: 'integration',
+          id: ctx.integrationId,
+          name: 'activeSyncLock',
+        })
+      } catch (error) {
+        syncLock = null
+      }
+
+      const lockPayload = syncLock?.state?.payload as { runId: string; timestamp: number; offset: number } | undefined
+      const now = Date.now()
+
+      // 5 minute sync lock timeout
+      const SYNC_LOCK_TIMEOUT = 5 * 60 * 1000
+
+      // if another sync is active for this run within the last 5 minutes, skip this duplicate
+      if (lockPayload?.runId === runId && now - lockPayload.timestamp < SYNC_LOCK_TIMEOUT) {
+        logger
+          .forBot()
+          .info(`⏭️ Sync already in progress for ${runId} at offset ${lockPayload.offset}, skipping duplicate webhook`)
+        return
+      }
+    }
+
+    // acquire lock before processing to prevent parallel execution
```

Comment:
question: there's a lot of logic related to making sure we don't run concurrently, could it be extracted easily?